core/federated/RTI/rti_common.h

```diff
@@ -53,17 +53,20 @@ typedef struct scheduling_node_t {
   tag_t last_granted;               // The maximum TAG that has been granted so far (or NEVER if none granted)
   tag_t last_provisionally_granted; // The maximum PTAG that has been provisionally granted (or NEVER if none granted)
   tag_t next_event;                 // Most recent NET received from the scheduling node (or NEVER if none received).
+  tag_t last_DNET;                  // Most recent DNET.
   scheduling_node_state_t state;    // State of the scheduling node.
-  uint16_t* upstream;               // Array of upstream scheduling node ids.
-  interval_t* upstream_delay;       // Minimum delay on connections from upstream scheduling nodes.
-                                    // Here, NEVER encodes no delay. 0LL is a microstep delay.
-  int num_upstream;                 // Size of the array of upstream scheduling nodes and delays.
-  uint16_t* downstream;             // Array of downstream scheduling node ids.
-  int num_downstream;               // Size of the array of downstream scheduling nodes.
-  execution_mode_t mode;            // FAST or REALTIME.
-  minimum_delay_t* min_delays;      // Array of minimum delays from upstream nodes, not including this node.
-  size_t num_min_delays;            // Size of min_delays array.
-  int flags;                        // Or of IS_IN_ZERO_DELAY_CYCLE, IS_IN_CYCLE
+  uint16_t* immediate_upstreams;    // Array of immediate upstream scheduling node ids.
+  interval_t* immediate_upstream_delays; // Minimum delay on connections from immdediate upstream scheduling nodes.
+                                         // Here, NEVER encodes no delay. 0LL is a microstep delay.
+  uint16_t num_immediate_upstreams;      // Size of the array of immediate upstream scheduling nodes and delays.
+  uint16_t* immediate_downstreams;       // Array of immediate downstream scheduling node ids.
+  uint16_t num_immediate_downstreams;    // Size of the array of immediate downstream scheduling nodes.
+  execution_mode_t mode;                 // FAST or REALTIME.
+  uint16_t* all_upstreams;               // Array of all upstream scheduling node ids.
+  uint16_t num_all_upstreams;            // Size of the array of all upstream scheduling nodes and delays.
+  uint16_t* all_downstreams;             // Array of all downstream scheduling node ids.
+  uint16_t num_all_downstreams;          // Size of the array of all downstream scheduling nodes.
```

Comment:
```suggestion
  uint16_t* immediate_upstream;    // Array of immediate upstream scheduling node ids.
  interval_t* immediate_upstream_delays; // Minimum delay on connections from immdediate upstream scheduling nodes.
                                         // Here, NEVER encodes no delay. 0LL is a microstep delay.
  uint16_t num_immediate_upstream;      // Size of the array of immediate upstream scheduling nodes and delays.
  uint16_t* immediate_downstream;       // Array of immediate downstream scheduling node ids.
  uint16_t num_immediate_downstream;    // Size of the array of immediate downstream scheduling nodes.
  execution_mode_t mode;                 // FAST or REALTIME.
  uint16_t* all_upstream;               // Array of all upstream scheduling node ids.
  uint16_t num_all_upstream;            // Size of the array of all upstream scheduling nodes and delays.
  uint16_t* all_downstream;             // Array of all downstream scheduling node ids.
  uint16_t num_all_downstream;          // Size of the array of all downstream scheduling nodes.
```

Reply:
Now I'm thinking we don't really need the `all_upstream` and `all_downstream` fields at all.  Once you have constructed the `min_delays` array, you can iterate over all upstream from ID `i` as follows (pseudo code, where `n` is the number of nodes):
```
  for (j from 0 to n-1) {
    if (min_delays[i][j] != FOREVER_TAG && i != j) {
      ... found a downstream node
    }
  }
```
To iterate over all upstream nodes, just reverse `i` and `j` in the `if` statement above.

Reply:
Thanks for the suggestion, @edwardalee! I remove the fields `all_upstream` and `all_downstream` in [adc0a1d](https://github.com/lf-lang/reactor-c/pull/349/commits/adc0a1d4250bba7b29a39cb4dce635b0db7e5e0c). I resolved all other related comments, too.